test/integration/zap-mint.ts

```diff
@@ -0,0 +1,239 @@
+import { EXTERNAL_USD_STABLE } from "./../../utils/Constants";
+import { BigNumber } from "ethers";
+import hre from "hardhat";
+import { ZapMint } from "./../../typechain/ZapMint";
+import chai from "chai";
+import { solidity } from "ethereum-waffle";
+import cap from "chai-as-promised";
+import {
+  getAllBDStablePools,
+  getBdEu,
+  getBDStableWethPool,
+  getBdUs,
+  getBdx,
+  getDeployer,
+  getERC20,
+  getTreasurySigner,
+  getUniswapRouter,
+  getWbtc,
+  getWeth
+} from "../../utils/DeployedContractsHelpers";
+import { numberToBigNumberFixed, to_d18 } from "../../utils/NumbersHelpers";
+import { setUpFunctionalSystemForTests } from "../../utils/SystemSetup";
+import { generateAllPaths, getAllTokensAddresses } from "../../utils/UniswapPoolsHelpers";
+import { IERC20 } from "../../typechain/IERC20";
+import { BdStablePool } from "../../typechain/BdStablePool";
+import { expectToFail } from "../helpers/common";
+
+chai.use(cap);
+
+chai.use(solidity);
+const { expect } = chai;
+
+//todo
+//add test for collateral ratio < 100% (mint which use bdx)
+//use one initialized zapmint object for all tests..? (alternatively - move initialization to private method)
+//move common variables to constants (like deadline, router)
+//add test for zap minting directly from ETH/WBTC (ETH/WBTC in path[0], path.length = 1)
+//add tests for all require statements
+//add test for reentrancy?
+describe("Zap mint", () => {
+  beforeEach(async () => {
+    await hre.deployments.fixture();
+    const deployer = await getDeployer(hre);
+    await setUpFunctionalSystemForTests(hre, 1);
+    const treasury = await getTreasurySigner(hre);
+    const bdx = await getBdx(hre);
+    const weth = await getWeth(hre);
+    const wbtc = await getWbtc(hre);
+    const allInputTokens = await getAllTokensAddresses(hre, [bdx.address, weth.address, wbtc.address]);
+    for (const token of allInputTokens) {
+      const erc20 = await getERC20(hre, token);
+      const decimals = await erc20.decimals();
+      erc20.connect(treasury).transfer(deployer.address, numberToBigNumberFixed(10, decimals));
+    }
+  });
+
+  it("Should zap mint for every possible non-singular path", async () => {
+    const deployer = await getDeployer(hre);
+    const bdx = await getBdx(hre);
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
```

Comment:
FYI current block changes between operations

Reply:
I only calculate deadline based on that (th deadline is currentblock +10000 so that should be enough deadline)